src/vision/status_quo/barbara_polls_a_mutex.md

```diff
@@ -0,0 +1,151 @@
+potentially be futures underneath.
+
+Barbara wants to store a namespace which maps variable names to their values. She
+chooses to use a `HashMap` and finds the `async_lock` crate provides an async
+mutex, which she can use for concurrency. She determines she'll need a lock
+around the namespace itself to protect concurrent modification.
+
+For the entries in her map, Barbara decides to implement a two-variant enum. One
+variant indicates that there is no implicit asynchronicity to resolve and the
+value is stored directly here. The other variant indicates that this value is
+being computed asynchronously and polling will be required to resolve it.
+Because an asynchronous task might want to change one of these entries from the
+asynchronous variant to the ready variant, she'll need to wrap the entries in
+an `Arc` and a `Mutex` to allow an asynchronous task to update them.
+
+Barbara wants to be able to derive a future from the entries in her namespace
+that will allow her to wait until the entry becomes ready and read the value.
+She decides to implement the `Future` trait directly. She's done this before
+for a few simple cases, and is somewhat comfortable with the idea, but she runs
+into significant trouble trying to deal with the mutex in the body of her poll
+function. Here are her attempts:
+
+```rust=
+use async_lock::Mutex;
+
+enum Value {
+    Int(i32),
+}
+
+enum NSEntry {
+    Ready(Value),
+    Waiting(Vec<Waker>),
+}
+
+type Namespace = Mutex<String, Arc<Mutex<NSEntry>>>;
+
+// Attempt 1: This compiles!!
+struct NSValueFuture(Arc<Mutex<NSEntry>>);
+impl Future for NSValueFuture {
+    type Output = Value;
+    pub fn poll(
+        self: Pin<&mut Self>, 
+        cx: &mut Context<'_>
+    ) -> Poll<Self::Output> {
+        let entry = match self.0.lock().poll() {
+            Poll::Ready(ent) => ent,
+
+            // When this returns, it will drop the future created by lock(),
+            // which drops our position in the lock's queue.
+            // You could never wake up.
+            // Get starved under contention. / Destroy fairness properties of lock.
+            Poll::Pending => return Poll::Pending,
+        };
+
+        ...
+    }
+}
+
+// Attempt 2
+struct NSValueFuture {
+    ent: Arc<Mutex<NSEntry>>,
+    lock_fut: Option<MutexGuard<'_, NSEntry>>,
+}
+impl Future for NSValueFuture {
+    type Output = Value;
+    pub fn poll(
+        self: Pin<&mut Self>, 
+        cx: &mut Context<'_>
+    ) -> Poll<Self::Output> {
+        if self.lock_fut.is_none() {
+            self.lock_fut = Some(self.ent.lock()),
+        }
+        // match self.lock_fut.unwrap().poll(cx)
+        // Pulled out pin-project, got confused, decided to just use unsafe.
+        match unsafe { Pin::new_unchecked(&mut self).lock_fut.unwrap() }.poll(cx) {
+            ...
+        }
+        // ??? lifetime for MutexLockFuture ???
+        // try async-std, async-lock
+    }
+}
+
+// Realize `lock_arc()` is a thing
```

Comment:
I'm lost here. Perhaps some more exposition can take us through Barbara's thought process?

Reply:
I seem to recall that the necessary property is that [`MutexGuardArc<T>`](https://docs.rs/async-std/1.9.0/async_std/sync/struct.MutexGuardArc.html) impls `Send` if `T` is `Send`, unlike other `MutexGuard` types.